Why SQL for Infrastructure Management (#2125)
iasql/iasql: site/blog/updates/why-sql-for-infrastructure.mdx

<!--truncate-->

## Following these instructions is imperative

The simplest way to define how to set something up is to write up a list of instructions to follow, in order, to build whatever infrastructure you're dealing with, whether its the instructions on how to build a restaurant or an AWS Fargate cluster. This list of steps to process (with a LISt Processor, or [LISP](https://en.wikipedia.org/wiki/Lisp_%28programming_language%29), if you will 😉), which can include instructions to repeat or skip over steps based on conditions you have run into, is called [imperative programming](https://en.wikipedia.org/wiki/Imperative_programming) in the software world.

:::note Imperative Restaurant Instructions

![Man in hard-hat builds your restaurant](/img/why-sql-for-infrastructure/man-in-hard-hat-builds-your-restaurant.png)

1. Design restaurant.
2. Flatten the ground.
3. Prop up pieces of wood next to each other.
4. Bang nails into wood.
5. Is restaurant finished? No, go to 3.

:::

> For your cloud infrastructure, this is similar to using the [AWS SDK](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/) and directly calling the various methods, checking their results and eventually arriving at the desired infrastructure. It is *also* like following a step-by-step guide clicking through the various AWS console UI elements to set up your infrastructure, like in [this guide](https://towardsdatascience.com/deploy-your-python-app-with-aws-fargate-tutorial-7a48535da586). The latter may not be automated, to you, but to your company executives it is as they don't care how the infrastructure building was accomplished as long as it was done quickly, cheaply, and reliably, and you know they always want all three. 😉

Author: Do you think we can include it without distracting? I never used ansible, so I wouldn't want to undersell/underdescribe it.

> For your cloud infrastructure, this is similar to using the [AWS SDK](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/) and directly calling the various methods, checking their results and eventually arriving at the desired infrastructure. It is *also* like following a step-by-step guide clicking through the various AWS console UI elements to set up your infrastructure, like in [this guide](https://towardsdatascience.com/deploy-your-python-app-with-aws-fargate-tutorial-7a48535da586). The latter may not be automated, to you, but to your company executives it is as they don't care how the infrastructure building was accomplished as long as it was done quickly, cheaply, and reliably, and you know they always want all three. 😉

Reviewer: I think it would be too much. That can come in another blog post about the `ssh-*` modules and we can point back to this one

:::note Declarative Restaurant Instructions

![King in robes on throne](/img/why-sql-for-infrastructure/king-in-robes-on-throne.png)

*By decree of the king, a restaurant shall be built!*

:::

In declarative programming you have some initial state and in comparatively dense and high-level declarative code define the desired state. In many use-cases (like web browsers and sometimes for restaurant-building contractors) the initial state is "blank" and the engine that transforms the declarative code into imperative operations can often be a relatively straightforward [interpreter](https://en.wikipedia.org/wiki/Interpreter_%28computing%29) walking the [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) of the declarative code.

> Infrastructure management *generally* is not that straightforward. Changes to infrastructure require in-place mutations of the current state. A full [blue/green deployment](https://en.wikipedia.org/wiki/Blue-green_deployment) of the entirety of your production infrastructure is very expensive, requiring all resource costs to be doubled, and deployments to require some amount of downtime for users to swap over and state to be resynchronized. Particularly for databases this approach is essentially impossible as there is too much state to swap.

Reviewer: ```suggestion
Infrastructure management *generally* is not that straightforward. Changes to infrastructure require in-place mutations of the current state, or a full [blue/green deployment](https://en.wikipedia.org/wiki/Blue-green_deployment) of the entirety of your production infrastructure is very expensive, requiring all resource costs to be doubled, and deployments to require some amount of downtime for users to swap over and state to be resynchronized. Particularly for databases this approach is essentially impossible as there is too much state to swap which is why database migrations are a nightmare.
```

Author: I don't know if I want to say they're a nightmare? I didn't really have much issue with them at CarDash. It gets hard when there are concurrent migration efforts happening by different engineers, which is what I think caused issues back in the early days at Uber.

I'd probably tone it down a bit to:

```suggestion
Infrastructure management *generally* is not that straightforward. Changes to infrastructure require in-place mutations of the current state or a full [blue/green deployment](https://en.wikipedia.org/wiki/Blue-green_deployment) of the entirety of your production infrastructure is very expensive, requiring all resource costs to be doubled, and deployments to require some amount of downtime for users to swap over and state to be resynchronized. Particularly for databases this approach is essentially impossible as there is too much state to swap, which is what makes database migrations tricky when dropping data.
```

Needing to learn Terraform's HCL language to make infrastructure changes has been cited as a [weakness](https://www.pulumi.com/why-pulumi/) that can be mitigated by wrapping it in the syntaxes more familiar to the user like [Terraform CDK](https://developer.hashicorp.com/terraform/cdktf), [Pulumi](https://www.pulumi.com), or [AWS CDK](https://docs.aws.amazon.com/cdk/v2/guide/home.html). This *is* a weakness, but we believe that the HCL-to-Statefile transformation being [surjective](https://en.wikipedia.org/wiki/Surjective_function) instead of [bijective](https://en.wikipedia.org/wiki/Bijection) is actually the bigger issue.

> 

Reviewer: let's include a surjective diagram analogous to the bijective one in the next section

Author: :::note Surjective Function *f* Diagram

![Surjective Diagram](https://upload.wikimedia.org/wikipedia/commons/6/6c/Surjection.svg)

[*From Wikipedia*](https://en.wikipedia.org/wiki/Surjective_function#/media/File:Surjection.svg) A function *f* is surjective if all of its inputs *X* map to all outputs *Y*, but overlaps can occur.

This is like multiple Terraform HCL representations all mapping to the exact same cloud infrastructure.

<!-- TODO: This section could ideally show some way you can trigger a corruption of the statefile vs AWS's actual state -->

## The Objective is Bi- hmmmm....

:::note Bijective Function *f* Diagram

![Bijective Diagram](https://upload.wikimedia.org/wikipedia/commons/a/a5/Bijection.svg)

[*From Wikipedia*](https://en.wikipedia.org/wiki/Bijection#/media/File:Bijection.svg) A function *f* is bijective if all of its inputs *X* map to all outputs *Y* exactly once.

:::

A bijective function has a one-to-one mapping of all state in one set with all state in another set. Both representations are equivalent (given the appropriate transform functions), both sets (of possible cloud states) are the same size (conceptually), and for any one particular state in one set there is exactly one state in the other set. Programming Languages don't fit the bill here, as was demonstrated earlier, but what does? What, exactly, defines your cloud resources in the most minimal way?

> The answer is ~~[water vapor](https://en.wikipedia.org/wiki/Cloud)~~ [Entities](https://en.wikipedia.org/wiki/Entity#In_computer_science). Each cloud resource has some unique identifier and a set of properties that configure it. Some of those properties can be relations or dependencies on other cloud resources. The ordering of the entities does not matter, so long as the dependencies are explicitly defined to make determination of the order of operations possible. You may have noticed that this sounds a lot like [Objects](https://en.wikipedia.org/wiki/Object_%28computer_science%29) in [Object-Oriented Programming](https://en.wikipedia.org/wiki/Object-oriented_programming). It is very much the same, with the only difference being that the unique identifier *must* be unique, so if we considered an array of pointers to these objects as the entity IDs, multiple pointers to the same object wouldn't be allowed.

Reviewer: ```suggestion
The best answer in most cases is ~~[water vapor](https://en.wikipedia.org/wiki/Cloud)~~ [Entities](https://en.wikipedia.org/wiki/Entity#In_computer_science). Each cloud resource has some unique identifier and a set of properties that configure it. Some of those properties can be relations or dependencies on other cloud resources. The ordering of the entities does not matter, so long as the dependencies are explicitly defined to make determination of the order of operations possible. You may have noticed that this sounds a lot like [Objects](https://en.wikipedia.org/wiki/Object_%28computer_science%29) in [Object-Oriented Programming](https://en.wikipedia.org/wiki/Object-oriented_programming). It is very much the same, with the only difference being that the unique identifier *must* be unique, so if we considered an array of pointers to these objects as the entity IDs, multiple pointers to the same object wouldn't be allowed.
```

make it less sales-y?

SQL immediately stands out here because it was designed for making [relational algebra](https://en.wikipedia.org/wiki/Relational_algebra), the other side of the [Entity-Relationship model](https://en.wikipedia.org/wiki/Entity%E2%80%93relationship_model), accessible. There are likely more people who know SQL than any programming language (for IaC) or data format you could choose to represent your cloud infrastructure. Many non-programmers know it, as well, such as data scientists, business analysts, accountants, etc, and there is an entire ecosystem of useful functionality you gain "for free" by using a SQL representation than just about any format out there. Furthermore, a powerful SQL engine like the open source [PostgreSQL](https://postgresql.org) provides tools capable of elevating the infrastructure management *experience* beyond what IaC can do today.

> ### The Advantages of SQL Databases for Infrastructure Management

Reviewer: ```suggestion
## The Advantages of SQL Databases for Infrastructure Management
```

Reviewer: ?

Author: Why this suggestion? I considered it a sub-section of the section it's currently in.

First, by defining [foreign key relations](https://www.postgresql.org/docs/current/tutorial-fk.html) between tables, the database can immediately inform you if you have dependencies that must be created before you can create the resource you desire. It will simply return an error message to you if you have inserted an incomplete record missing any required fields.

> "But," you say, "programming languages with solid type systems, like Rust, can also do that for you." That is true, and this was not an attempt to disparage them, merely to show that you aren't losing anything by using SQL as the representation. There are, however, ways SQL goes above and beyond. By defining [unique, not-null, and check constraints](https://www.postgresql.org/docs/current/ddl-constraints.html) on top of a [rich type system](https://www.postgresql.org/docs/current/datatype.html) the database can prevent you from inserting faulty data to a greater degree than even [Rust](https://www.rust-lang.org), as the constraints can be against the live state of your infrastructure itself. Instead of code needing to deal with unexpected values further down the line, the faulty data can be prevented from ever being inserted. Rust's compiler making sure that you handle every possible branch is an awesome thing, but SQL has the power to make more of these branches impossible in the first place.

Reviewer 2: nit:
```suggestion
"But," you say, "programming languages with solid type systems, like [Rust](https://www.rust-lang.org), can also do that for you." That is true, and this was not an attempt to disparage them, merely to show that you aren't losing anything by using SQL as the representation. There are, however, ways SQL goes above and beyond. By defining [unique, not-null, and check constraints](https://www.postgresql.org/docs/current/ddl-constraints.html) on top of a [rich type system](https://www.postgresql.org/docs/current/datatype.html) the database can prevent you from inserting faulty data to a greater degree than even Rust, as the constraints can be against the live state of your infrastructure itself. Instead of code needing to deal with unexpected values further down the line, the faulty data can be prevented from ever being inserted. Rust's compiler making sure that you handle every possible branch is an awesome thing, but SQL has the power to make more of these branches impossible in the first place.
```

SQL Database [Snapshotting](https://en.wikipedia.org/wiki/Snapshot_%28computer_storage%29) could be used to quickly revert all infrastructure changes for infrastructure without hidden state (eg, the snapshot of your load balancers' prior state would successfully restore the load balancer configuration, but a dropped database restored via a snapshot would not restore the data within, only that you had such a database in the first place and you would still need to manually re-load a backup snapshot of those, as well).

Database engines also represent a standardized protocol to interface with your data. This standard opens up a world of possibilities that can encompass the imperative and declarative infrastructure management styles and go beyond them. You can:

>   - Perform operations on your database via scripts and the [`psql` CLI](https://www.postgresql.org/docs/current/app-psql.html) tool, using whatever mix of declarative and imperative styles you prefer, mutating various tables declaratively and calling [`AWS SDK functions directly`](/docs/modules/aws/aws_sdk/?view=code-examples) where and when you see fit.

Reviewer: I would mention that declarative is possible, but more cumbersome in SQL

Author: > using whatever mix of declarative and imperative styles you prefer

This doesn't cover that?

Author: It's also from the perspective of "the DB records themselves are the declarative 'code'" not "you can make idempotent SQL queries to make the query itself quasi-declarative" which I didn't touch on.

Reviewer: I see. I guess I don't think about it in the first way that you describe it 🤔

Peer review of changes is *different* versus IaC. Instead of writing HCL or other IaC code, having it reviewed, and then `apply`ing it, you can enter an [IaSQL transaction](/docs/transaction), make the desired changes with Postgres acting as your IDE, automatically blocking many impossible changes via types and relations, and use [`iasql_create_review`](/docs/review/) to create a review artifact for your team *after* IaSQL has already vetted it for technical correctness. This vetting makes the result superior to the traditional peer review system in IaC.

Database schemas can be overwhelming, and a schema accurately representing *all* of AWS would be dizzying. HCL and other IaC tools let you elide parts of AWS via a module system, so we have implemented our own [module system](/docs/module) on top of Postgres inspired equally by programming language module systems and Linux distro package managers. This module system takes the place the usual database [schema migration system](https://en.wikipedia.org/wiki/Schema_migration) for the management of the tables in the database, which may feel unusual at first, but we believe it is superior (for this use-case).

> Even with IaC tools letting you ignore details you don't specifically need, AWS and other clouds expose many details in multiple layers for you to control when often you just need the same subset of functionality over-and-over, so these IaC tools have higher-level modules for these cookie-cutter use-cases. For IaSQL we [did the same](/docs/low-level-vs-high-level), though in a way that coexists with the lower-level modules so you can still query those details later, if you like.

Reviewer: ```suggestion
Even with IaC tools letting you ignore details you don't specifically need, AWS and other clouds expose many details in multiple layers for you to control when often you just need the same subset of functionality over-and-over, so these IaC tools have higher-level modules for these cookie-cutter use-cases. For IaSQL we [did the same](/docs/low-level-vs-high-level), though in a way that coexists with the lower-level modules so you can still query those details later, if you like. [Here](/blog/ecs-simplified) is an example of a high-level module that greatly simplifies ECS fargate. 
```

INSERT INTO instance (ami, instance_type, key_pair_name, state, tags, hibernation_enabled, region, subnet_id)
VALUES ('resolve:ssm:/aws/service/canonical/ubuntu/server/20.04/stable/current/amd64/hvm/ebs-gp2/ami-id', 't3.small', 'login', 'running', '{"name":"login"}'::jsonb, 'f', (SELECT region FROM aws_regions WHERE region = 'us-west-2'), (SELECT id FROM subnet WHERE subnet_id = 'subnet-06140fd708a495450' AND region = (SELECT region FROM aws_regions WHERE region = 'us-west-2')));

INSERT INTO instance_security_groups (instance_id, security_group_id)
VALUES ((SELECT id FROM instance WHERE instance_id = NULL AND region = (SELECT region FROM aws_regions WHERE region = 'us-west-2')), (SELECT id FROM security_group WHERE group_id = NULL AND region = (SELECT region FROM aws_regions WHERE region = 'us-west-2')));
```

which in this case is very similar to (though a bit more verbose than) our original SQL statements, but if you were going back and forth on what exact changes you're going to make, it can be a good summary.

> If we're good with this, we can `SELECT iasql_apply()` to push this to our cloud account and then update our `ssh_credentials` record with the public IP address and check the connection:

Reviewer 2: ```suggestion
If we're good with this, we can `SELECT iasql_commit()` to push this to our cloud account and then update our `ssh_credentials` record with the public IP address and check the connection:
```

```sql
WITH h AS (
  SELECT host(im.public_ip_address) AS hostname
  FROM instance_metadata im
  INNER JOIN instance i ON im.instance_id = i.instance_id
  WHERE i.tags ->> 'name' = 'login'
  LIMIT 1
)
UPDATE ssh_credentials SET hostname = h.hostname FROM ssh_credentials, h WHERE hostname = 'tbd';

SELECT * FROM ssh_exec('login', 'uname -a');
```

> ## SQL is the best option for Infrastructure Management

Reviewer: ```suggestion
## SQL as the best of the worst option for Infrastructure Management
```

a bit nihilistic or cynical, but keeps the tone from the beginning?

Author: "the best of the worst" reads a bit weird to me. I think this would work a bit better?

```suggestion
## SQL is the least bad option for Infrastructure Management
```

Reviewer: that is better!

Data, on the other hand, is much easier to normalize in such a way, and relational data models, as found in SQL databases, can naturally model the relationships between cloud services, making it possible to produce a bijective representation. IaSQL takes advantage of this to make a declarative model based on tables and foreign keys that can automatically handle changes done outside of this declarative model, either by dropping down into its own imperative mode or by using any other tooling to manage your infrastructure, giving it a flexibility and resiliency that IaC tools don't have.

On top of this foundation, tooling to replicate or substitute for existing IaC tooling has been added, along with all of the past ~50 years of database tooling and expertise that you can take advantage of, providing a superset of functionality that makes existing use-cases easier and new use-cases possible.

> IaSQL is [currently in beta](/blog/beta) and can be used locally with [`docker`](https://hub.docker.com/r/iasql/iasql) or via our [SaaS](https://app.iasql.com). We're proud of what we have built, and can't wait to hear from you on feature requests and bug reports.

Reviewer: ```suggestion
IaSQL is [currently in beta](/blog/beta) and can be used locally with [`docker`](https://hub.docker.com/r/iasql/iasql) or via our [SaaS](/hosted). We're proud of what we have built, and can't wait to hear from you on feature requests and bug reports.
```

```hcl
terraform {
  required_providers {
    aws = {
      source = "hashicorp/aws"
      version = "4.60.0"
    }
  }
}

data "external" "example" {
  program = [
    "bash",
    "-c",
    <<-EOT
      if [ ! -f ./login.pem ]; then
        # Terraform doesn't support creating the PEM file via AWS. It has to be
        # created locally and then imported into AWS. Inside of a conditional so
        # we don't accidentally re-create the file between 'plan' and 'apply'
        yes '' | ssh-keygen -t ed25519 -m PEM -f login.pem
      fi
      chmod 600 login.pem
      # The output of this script needs to be JSON formatted. This is a bit wonky
      # but building it over time via 'jq' operators is actually harder to follow
      echo {
      echo   '"access_key_id": "'$(aws configure get aws_access_key_id)'",'
      echo   '"secret_access_key": "'$(aws configure get aws_secret_access_key)'",'

>       echo   '"public_key": "'$(cat login.pem.pub)'"'

Reviewer 3: I think Terraform will use the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables. Also, for the key we can use [`local_file`](https://registry.terraform.io/providers/hashicorp/local/latest/docs/data-sources/file) instead?

Author: Yes, but I didn't like the not-self-contained nature of that approach versus this one for both situations. It requires us to invoke `terraform plan` like `AWS_ACCESS_KEY_ID=$(aws configure get aws_access_key_id) AWS_SECRET_ACCESS_KEY=$(aws configure get aws_secret_access_key) terraform plan` if we don't want to make the assumption that those environment variables were already set.

Also, I hate environment variables because they are common footguns when trying to reproduce what some piece of code does, and I prefer to minimize them as much as possible.

Author: I can add a comment into the Terraform code that the shell script isn't necessary if you set the appropriate environment variables outside and create the ssh key ahead of time and just reference it here, but those would be manual steps you have to ensure you do before running the rest of the Terraform code which makes actually executing the example more difficult.

}
```

This example still requires the following lines from the imperative example to get the public IP address of the new EC2 instance and log into it:

```bash

> PUBLIC_IP=$(aws ec2 describe-instances \

Reviewer 3: Did you consider making this an [HCL output](https://developer.hashicorp.com/terraform/language/values/outputs) and [accessing it via bash](https://stackoverflow.com/a/69260021/2594912)?

Author: So, the terraform code was the longest of the three, so I thought adding that and then still having a bash script afterwards would unfairly paint Terraform even worse, but I suppose an argument could be made that I'm unfairly painting Terraform as being unable to get that public IP address to use in the bash script.

I *know* the Terraform code is going to draw rotten tomatoes at us either way, but let me think about which is "worse." (And go with that one? :zany_face:)
